File: src/journal/tests/testjournal.cpp
#include "doctest.h"
#include "fakeit.hpp"
#include "..\ioutput.h"
#include "..\iconfig.h"
#include "..\application.h"
#include "..\ishell.h"
#include <sstream>

using namespace fakeit;

TEST_CASE("Do nothing") {
    Mock<IConfig> configMock;
    Mock<IShell> shellMock;
    Mock<IOutput> blogMock;
    Mock<IOutput> bookMock;
    std::ostringstream os;

    Fake(Method(configMock, get_basedir));
    Fake(Method(configMock, get_scriptdir));
    Fake(Method(configMock, is_blog));
    Fake(Method(configMock, is_book));
    Fake(Method(configMock, is_publish));
    Fake(Method(configMock, is_backup));
    Fake(Method(configMock, is_git_commit_push_private_repo));
    Fake(Method(configMock, get_private_dir));
    Fake(Method(shellMock, setup_encoding));
    Fake(Method(shellMock, create_backup));
    Fake(Method(shellMock, git_commit_push));
    Fake(Method(shellMock, current_path));
    When(Method(shellMock, cout)).AlwaysReturn(os);
    Fake(Method(blogMock, generate));
    Fake(Method(bookMock, generate));

    Application sut(blogMock.get(), bookMock.get(), configMock.get(), shellMock.get());
    sut.run();
}

TEST_CASE("Create blog") {
    Mock<IConfig> configMock;
    Mock<IShell> shellMock;
    Mock<IOutput> blogMock;
    Mock<IOutput> bookMock;
    std::ostringstream os;

    Fake(Method(configMock, get_basedir));
    Fake(Method(configMock, get_scriptdir));
    When(Method(configMock, is_blog)).AlwaysReturn(true);
    Fake(Method(configMock, is_book));
    Fake(Method(configMock, is_publish));
    Fake(Method(configMock, is_backup));
    Fake(Method(configMock, is_git_commit_push_private_repo));
    Fake(Method(configMock, get_private_dir));
    Fake(Method(shellMock, setup_encoding));
    Fake(Method(shellMock, create_backup));
    Fake(Method(shellMock, git_commit_push));
    Fake(Method(shellMock, current_path));
    When(Method(shellMock, cout)).AlwaysReturn(os);
    Fake(Method(blogMock, generate));
    Fake(Method(bookMock, generate));

    Application sut(blogMock.get(), bookMock.get(), configMock.get(), shellMock.get());
    sut.run();

    Verify(Method(shellMock, create_backup)).Never();
    Verify(Method(shellMock, git_commit_push)).Never();
    Verify(Method(blogMock, generate)).Once();
    Verify(Method(bookMock, generate)).Never();
}

TEST_CASE("Create book") {
    Mock<IConfig> configMock;
    Mock<IShell> shellMock;
    Mock<IOutput> blogMock;
    Mock<IOutput> bookMock;
    std::ostringstream os;

    Fake(Method(configMock, get_basedir));
    Fake(Method(configMock, get_scriptdir));
    Fake(Method(configMock, is_blog));
    When(Method(configMock, is_book)).AlwaysReturn(true);
    Fake(Method(configMock, is_publish));
    Fake(Method(configMock, is_backup));
    Fake(Method(configMock, is_git_commit_push_private_repo));
    Fake(Method(configMock, get_private_dir));
    Fake(Method(shellMock, setup_encoding));
    Fake(Method(shellMock, create_backup));
    Fake(Method(shellMock, git_commit_push));
    Fake(Method(shellMock, current_path));
    When(Method(shellMock, cout)).AlwaysReturn(os);
    Fake(Method(blogMock, generate));
    Fake(Method(bookMock, generate));

    Application sut(blogMock.get(), bookMock.get(), configMock.get(), shellMock.get());
    sut.run();

    Verify(Method(shellMock, create_backup)).Never();
    Verify(Method(shellMock, git_commit_push)).Never();
    Verify(Method(blogMock, generate)).Never();
    Verify(Method(bookMock, generate)).Once();
}

TEST_CASE("Publish blog") {
    Mock<IConfig> configMock;
    Mock<IShell> shellMock;
    Mock<IOutput> blogMock;
    Mock<IOutput> bookMock;
    std::ostringstream os;

    Fake(Method(configMock, get_basedir));
    Fake(Method(configMock, get_scriptdir));
    When(Method(configMock, is_blog)).AlwaysReturn(true);
    Fake(Method(configMock, is_book));
    When(Method(configMock, is_publish)).AlwaysReturn(true);
    Fake(Method(configMock, is_backup));
    Fake(Method(configMock, is_git_commit_push_private_repo));
    Fake(Method(configMock, get_private_dir));
    Fake(Method(shellMock, setup_encoding));
    Fake(Method(shellMock, create_backup));
    Fake(Method(shellMock, git_commit_push));
    Fake(Method(shellMock, current_path));
    When(Method(shellMock, cout)).AlwaysReturn(os);
    Fake(Method(blogMock, generate));
    Fake(Method(bookMock, generate));

    Application sut(blogMock.get(), bookMock.get(), configMock.get(), shellMock.get());
    sut.run();

    Verify(Method(shellMock, create_backup)).Never();
    Verify(Method(shellMock, git_commit_push)).Twice();
    Verify(Method(blogMock, generate)).Once();
    Verify(Method(bookMock, generate)).Never();
}

TEST_CASE("Publish blog and backup") {
    Mock<IConfig> configMock;
    Mock<IShell> shellMock;
    Mock<IOutput> blogMock;
    Mock<IOutput> bookMock;
    std::ostringstream os;

    Fake(Method(configMock, get_basedir));
    Fake(Method(configMock, get_scriptdir));
    When(Method(configMock, is_blog)).AlwaysReturn(true);
    Fake(Method(configMock, is_book));
    When(Method(configMock, is_publish)).AlwaysReturn(true);
    When(Method(configMock, is_backup)).AlwaysReturn(true);
    Fake(Method(configMock, is_git_commit_push_private_repo));
    Fake(Method(configMock, get_private_dir));
    Fake(Method(shellMock, setup_encoding));
    Fake(Method(shellMock, create_backup));
    Fake(Method(shellMock, git_commit_push));
    Fake(Method(shellMock, current_path));
    When(Method(shellMock, cout)).AlwaysReturn(os);
    Fake(Method(blogMock, generate));
    Fake(Method(bookMock, generate));

    Application sut(blogMock.get(), bookMock.get(), configMock.get(), shellMock.get());
    sut.run();

    Verify(Method(shellMock, create_backup)).Once();
    Verify(Method(shellMock, git_commit_push)).Twice();
    Verify(Method(blogMock, generate)).Once();
    Verify(Method(bookMock, generate)).Never();
}

TEST_CASE("Publish private notes") {
    Mock<IConfig> configMock;
    Mock<IShell> shellMock;
    Mock<IOutput> blogMock;
    Mock<IOutput> bookMock;
    std::ostringstream os;

    Fake(Method(configMock, get_basedir));
    Fake(Method(configMock, get_scriptdir));
    When(Method(configMock, is_blog)).AlwaysReturn(true);
    Fake(Method(configMock, is_book));
    When(Method(configMock, is_publish)).AlwaysReturn(true);
    When(Method(configMock, is_backup)).AlwaysReturn(true);
    When(Method(configMock, is_git_commit_push_private_repo)).AlwaysReturn(true);
    Fake(Method(configMock, get_private_dir));
    Fake(Method(shellMock, setup_encoding));
    Fake(Method(shellMock, create_backup));
    Fake(Method(shellMock, git_commit_push));
    Fake(Method(shellMock, current_path));
    When(Method(shellMock, cout)).AlwaysReturn(os);
    Fake(Method(blogMock, generate));
    Fake(Method(bookMock, generate));

    Application sut(blogMock.get(), bookMock.get(), configMock.get(), shellMock.get());
    sut.run();

    Verify(Method(shellMock, create_backup)).Once();
    Verify(Method(shellMock, git_commit_push)).Exactly(3_Times);
    Verify(Method(blogMock, generate)).Once();
    Verify(Method(bookMock, generate)).Never();
}

File: src/journal/dotnet/Journal.Test/TestJournal.cs
using Xunit;
using Journal;
using Journal.Interfaces;
using Moq;

namespace Journal.Tests;

public class TestJournal
{
    [Fact]
    public void TestJournalRunWithBlog()
    {
        // Arrange
        var mockBlog = new Mock<IBlog>();
        var mockBook = new Mock<IBook>();
        var mockConfig = new Mock<IConfig>();
        var mockShell = new Mock<IShell>();

        mockConfig.Setup(c => c.IsBlog()).Returns(true);
        mockConfig.Setup(c => c.IsBook()).Returns(false);
        mockConfig.Setup(c => c.IsPublish()).Returns(false);
        mockConfig.Setup(c => c.IsBackup()).Returns(false);
        mockConfig.Setup(c => c.IsGitCommitPushPrivateRepo()).Returns(false);
        mockConfig.Setup(c => c.GetBasedir()).Returns(Directory.GetCurrentDirectory());
        mockConfig.Setup(c => c.GetScriptdir()).Returns(Path.Combine(Directory.GetCurrentDirectory(), "src"));

        mockBlog.Setup(b => b.Create(It.IsAny<string>(), It.IsAny<string>())).Returns(0);

        var journal = new Journal(mockBlog.Object, mockBook.Object, mockConfig.Object, mockShell.Object);

        // Act
        journal.Run();

        // Assert
        mockShell.Verify(s => s.SetupEncoding(), Times.Once);
        mockShell.Verify(s => s.CurrentPath(It.IsAny<string>()), Times.Exactly(2)); // Once to change, once to restore
        mockBlog.Verify(b => b.Create(It.IsAny<string>(), It.IsAny<string>()), Times.Once);
        mockBook.Verify(b => b.Create(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<bool>()), Times.Never);
    }

    [Fact]
    public void TestJournalRunWithBook()
    {
        // Arrange
        var mockBlog = new Mock<IBlog>();
        var mockBook = new Mock<IBook>();
        var mockConfig = new Mock<IConfig>();
        var mockShell = new Mock<IShell>();

        mockConfig.Setup(c => c.IsBlog()).Returns(false);
        mockConfig.Setup(c => c.IsBook()).Returns(true);
        mockConfig.Setup(c => c.GetBasedir()).Returns(Directory.GetCurrentDirectory());
        mockConfig.Setup(c => c.GetScriptdir()).Returns(Path.Combine(Directory.GetCurrentDirectory(), "src"));

        mockBook.Setup(b => b.Create(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<bool>())).Returns(0);

        var journal = new Journal(mockBlog.Object, mockBook.Object, mockConfig.Object, mockShell.Object);

        // Act
        journal.Run();

        // Assert
        mockShell.Verify(s => s.SetupEncoding(), Times.Once);
        mockShell.Verify(s => s.CurrentPath(It.IsAny<string>()), Times.Exactly(2)); // Once to change, once to restore
        mockBlog.Verify(b => b.Create(It.IsAny<string>(), It.IsAny<string>()), Times.Never);
        mockBook.Verify(b => b.Create(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<bool>()), Times.Once);
    }

    [Fact]
    public void TestJournalRunWithBlogAndPublish()
    {
        // Arrange
        var mockBlog = new Mock<IBlog>();
        var mockBook = new Mock<IBook>();
        var mockConfig = new Mock<IConfig>();
        var mockShell = new Mock<IShell>();

        mockConfig.Setup(c => c.IsBlog()).Returns(true);
        mockConfig.Setup(c => c.IsBook()).Returns(false);
        mockConfig.Setup(c => c.IsPublish()).Returns(true);
        mockConfig.Setup(c => c.IsBackup()).Returns(false);
        mockConfig.Setup(c => c.IsGitCommitPushPrivateRepo()).Returns(false);
        mockConfig.Setup(c => c.GetBasedir()).Returns(Directory.GetCurrentDirectory());
        mockConfig.Setup(c => c.GetScriptdir()).Returns(Path.Combine(Directory.GetCurrentDirectory(), "src"));

        mockBlog.Setup(b => b.Create(It.IsAny<string>(), It.IsAny<string>())).Returns(0);

        var journal = new Journal(mockBlog.Object, mockBook.Object, mockConfig.Object, mockShell.Object);

        // Act
        journal.Run();

        // Assert
        mockShell.Verify(s => s.SetupEncoding(), Times.Once);
        mockShell.Verify(s => s.CurrentPath(It.IsAny<string>()), Times.Exactly(2)); // Once to change, once to restore
        mockBlog.Verify(b => b.Create(It.IsAny<string>(), It.IsAny<string>()), Times.Once);
        mockShell.Verify(s => s.GitCommitPush(It.IsAny<string>(), It.IsAny<string>()), Times.Exactly(2)); // Current dir and public/blog
    }

    [Fact]
    public void TestJournalRunWithBlogAndBackup()
    {
        // Arrange
        var mockBlog = new Mock<IBlog>();
        var mockBook = new Mock<IBook>();
        var mockConfig = new Mock<IConfig>();
        var mockShell = new Mock<IShell>();

        mockConfig.Setup(c => c.IsBlog()).Returns(true);
        mockConfig.Setup(c => c.IsBook()).Returns(false);
        mockConfig.Setup(c => c.IsPublish()).Returns(true);
        mockConfig.Setup(c => c.IsBackup()).Returns(true);
        mockConfig.Setup(c => c.IsGitCommitPushPrivateRepo()).Returns(false);
        mockConfig.Setup(c => c.GetBasedir()).Returns(Directory.GetCurrentDirectory());
        mockConfig.Setup(c => c.GetScriptdir()).Returns(Path.Combine(Directory.GetCurrentDirectory(), "src"));

        mockBlog.Setup(b => b.Create(It.IsAny<string>(), It.IsAny<string>())).Returns(0);

        var journal = new Journal(mockBlog.Object, mockBook.Object, mockConfig.Object, mockShell.Object);

        // Act
        journal.Run();

        // Assert
        mockShell.Verify(s => s.SetupEncoding(), Times.Once);
        mockShell.Verify(s => s.CreateBackup(It.IsAny<string>()), Times.Once);
        mockBlog.Verify(b => b.Create(It.IsAny<string>(), It.IsAny<string>()), Times.Once);
    }

    [Fact]
    public void TestJournalRunWithBlogAndPrivateRepo()
    {
        // Arrange
        var mockBlog = new Mock<IBlog>();
        var mockBook = new Mock<IBook>();
        var mockConfig = new Mock<IConfig>();
        var mockShell = new Mock<IShell>();

        mockConfig.Setup(c => c.IsBlog()).Returns(true);
        mockConfig.Setup(c => c.IsBook()).Returns(false);
        mockConfig.Setup(c => c.IsPublish()).Returns(false);
        mockConfig.Setup(c => c.IsBackup()).Returns(false);
        mockConfig.Setup(c => c.IsGitCommitPushPrivateRepo()).Returns(true);
        mockConfig.Setup(c => c.GetBasedir()).Returns(Directory.GetCurrentDirectory());
        mockConfig.Setup(c => c.GetScriptdir()).Returns(Path.Combine(Directory.GetCurrentDirectory(), "src"));
        mockConfig.Setup(c => c.GetPrivateDir()).Returns(Path.Combine(Directory.GetCurrentDirectory(), "private"));

        mockBlog.Setup(b => b.Create(It.IsAny<string>(), It.IsAny<string>())).Returns(0);

        var journal = new Journal(mockBlog.Object, mockBook.Object, mockConfig.Object, mockShell.Object);

        // Act
        journal.Run();

        // Assert
        mockShell.Verify(s => s.SetupEncoding(), Times.Once);
        mockShell.Verify(s => s.GitCommitPush(It.IsAny<string>(), "Add journal changes"), Times.Once);
        mockBlog.Verify(b => b.Create(It.IsAny<string>(), It.IsAny<string>()), Times.Once);
    }
} 